proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,199 @@
+experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The Chromium bug has a significant amount of developer interest.
+
+## Specification
+
+### Schema
+
+manifest.json
+```
+"icon_variants": [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png",
+  },
+  {
+    "16": "dark16.png",
+    "32": "dark32.png",
+    "color_scheme": "dark",
+  },
+  {
+    "16": "light16.png",
+    "32": "light32.png",
+    "color_scheme": ["dark", "light"]
+  }
+]
+```
+
+Set icon_variants dynamically.
+```
+const exampleProperties: {string: string | ImageData}[] = [
+  {
+    "any": "any.svg",
+  },
+  {
+    "16": "16.png",
+    "32": "32.png"
+  },
+  {
+    "16": darkImageData16,
+    "32": darkImageData32,
+    "color_scheme": "dark"
+  },
+  {
+    "16": lightImageData16,
+    "32": lightImageData32,
+    "color_scheme": "light"
+  }
+];
+
+// action.setIcon().
+const actionProperties = {variants: exampleProperties};
+action.setIcon(createProperties);
+
+// menus.*(), for supporting browsers.
+const menusProperties = {icon_variants: exampleProperties}
+menus.create(menusProperties);
+menus.update(menusProperties);
+```
+
+A benefit of this new structure is that it's more resiliant to future changes, thus allowing for more keys such as density (e.g. 2dppx), purpose (e.g.
+monochrome), and etc.
+
+## setIcon()
+Incumbent action.setIcon(), for reference.
+```
+action.setIcon({
+  path,
+  tabId
+  imageData,
+});
+```
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+1. If icon_variants are supplied, icons can be ignored.
+1. If icon_variants contain an icon object duplicate, the first match will be
+used.
+1. Any icon object that does not contain a "color_scheme" key will apply to both
+light and dark.
+1. If only one icon object is supplied, it will be used for both light and dark.
+1. icon_variants will not cause an error in the event that it is invalid
```

Comment:
Can we make this more specific? With the current wording, not recognising a single key in a single icon object would allow us to ignore the entire icon_variants key. I think we need to say that we will skip over individual icons in the array that contain a key we don't recognise.

Reply:
Done.